File: Data Structures/array/dart/array.dart
class CustomArray<T> {
  int length = 0;
  Map<int, T> _data = {};

  CustomArray();

  T? get(int index) {
    return _data[index];
  }

  int push(T item) {
    _data[length] = item;
    length++;
    return length;
  }

  T? pop() {
    if (length == 0) {
      return null;
    }
    T? lastItem = _data[length - 1];
    _data.remove(length - 1);
    length--;
    return lastItem;
  }

  T? deleteAt(int index) {
    T? item = _data[index];
    _shiftItems(index);
    return item;
  }

  int insertAt(int index, T item) {
    _shiftItemsForInsert(index);
    _data[index] = item;
    length++;
    return length;
  }

  void _shiftItems(int index) {
    for (int i = index; i < length - 1; i++) {
      _data[i] = _data[i + 1] as T;
    }
    _data.remove(length - 1);
    length--;
  }

  void _shiftItemsForInsert(int index) {
    for (int i = length; i > index; i--) {
      _data[i] = _data[i - 1] as T;
    }
  }

  void printArray() {
    List<T> elements = [];
    for (int i = 0; i < length; i++) {
      if (_data[i] != null) elements.add(_data[i]!);
    }
    print(elements);
  }
}

void main() {
  CustomArray<String> myArray = CustomArray<String>();
  myArray.push('hi');
  myArray.push('you');
  myArray.push('!');
  myArray.pop();
  myArray.deleteAt(0);
  myArray.push('there');
  myArray.insertAt(1, 'nice');
  myArray.printArray(); // Expect: [you, nice, there]
}

File: Data Structures/array/ts/array.ts
class CustomArray<T> {
  private length: number;
  private data: { [key: number]: T };

  constructor() {
    this.length = 0;
    this.data = {};
  }

  // Get element at index
  get(index: number): T {
    return this.data[index];
  }

  // Push element to end
  push(item: T): number {
    this.data[this.length] = item;
    this.length++;
    return this.length;
  }

  // Remove last element
  pop(): T {
    if (this.length === 0) {
      return undefined as any;
    }
    const lastItem = this.data[this.length - 1];
    delete this.data[this.length - 1];
    this.length--;
    return lastItem;
  }

  // Delete element at index
  delete(index: number): T {
    const item = this.data[index];
    this.shiftItems(index);
    return item;
  }

  // Insert element at index (manual implementation to show shift)
  insert(index: number, item: T): number {
    this.shiftItemsForInsert(index);
    this.data[index] = item;
    this.length++;
    return this.length;
  }

  private shiftItems(index: number) {
    for (let i = index; i < this.length - 1; i++) {
      this.data[i] = this.data[i + 1];
    }
    delete this.data[this.length - 1];
    this.length--;
  }

  private shiftItemsForInsert(index: number) {
    for (let i = this.length; i > index; i--) {
      this.data[i] = this.data[i - 1];
    }
  }

  print(): void {
    const elements = [];
    for (let i = 0; i < this.length; i++) {
      elements.push(this.data[i]);
    }
    console.log(`[${elements.join(", ")}]`);
  }
}

// Example Usage
const myArray = new CustomArray<string>();
myArray.push("hi");
myArray.push("you");
myArray.push("!");
myArray.pop();
myArray.delete(0);
myArray.push("there");
myArray.insert(1, "nice");
myArray.print(); // Expect: [you, nice, there]
